### ISSUE TITLE
render output with alpha

### ISSUE THREAD

**Issue by sfrankel** [google/filament#1441]
Hello,
Does Filament support render output with premultiplied alpha?  If so, can you please point me to an example?  Haven't found any in the samples dir.
Thanks!

**Comment by romainguy** [comment]
We have two such samples for Android and iOS:
https://github.com/google/filament/tree/master/android/samples/transparent-view
https://github.com/google/filament/tree/master/ios/samples/transparent-rendering

It's very easy, just pass the right flag to `Engine::createSwapChain()`:

```
engine->createSwapChain(nativeWindow, SwapChain::CONFIG_TRANSPARENT);
```

**Comment by sfrankel** [comment]
Great!  Thanks @romainguy I'll take a look

**Comment by thirdmenson** [comment]
Hey guys, Is there an example for windows on how to output with alpha channel? I modified the sample app to add the correct transparent flag and `mSwapChain = mFilamentApp->mEngine->createSwapChain(nativeWindow, SwapChain::CONFIG_TRANSPARENT); `  and I also added `view->setClearColor(math::float4(0.0f, 0.0f, 0.0f, 0.0f));` but the render output still doesn't have the alpha channel when I read the buffer using readPixel().

**Comment by romainguy** [comment]
@bejado

**Comment by thirdmenson** [comment]
I saw this comment in another issue. Looks like you don't support that on windows? has the transparent rendering support been added on windows? This is probably going to be make or break for the app I am working on.
 https://github.com/google/filament/issues/187#issuecomment-462822781

**Comment by bejado** [comment]
> render output still doesn't have the alpha channel when I read the buffer using readPixel()

Can you elaborate? How are you calling `readPixels` and what does the result alpha channel contain?

**Comment by thirdmenson** [comment]
This is how I am calling it:
```
Texture::PixelBufferDescriptor pb(image_data, area * 4, Texture::Format::RGBA, Texture::Type::UBYTE);
	renderer->readPixels(0, 0, winwidth, winheight, std::move(pb));
```

Sorry I should have said that the alpha channel value is there in the buffer but the value is always 255 for the entire image, and the background appears black on the resulting image where it should be transparent just like the other person is saying [here](https://github.com/google/filament/issues/187#issuecomment-462822781) . I have tried adding another layer behind the resulting image and check the transparency toggle in my app that I am copying the resulting image to and it's just black color.

**Comment by bejado** [comment]
I see, I'm able to reproduce. Let me dig a bit more into it, there's no reason I can think of why Windows wouldn't support transparency.

**Comment by thirdmenson** [comment]
Thank you so much for looking into this!

**Comment by bejado** [comment]
So there's actually a pretty simple fix for this. In [`PlatformWGL.cpp`](https://github.com/google/filament/blob/master/filament/backend/src/opengl/PlatformWGL.cpp#L92) set `cAlphaBits` on the pixel descriptor:

```
mPfd = {
    sizeof(PIXELFORMATDESCRIPTOR),
    1,
    PFD_DRAW_TO_WINDOW | PFD_SUPPORT_OPENGL | PFD_DOUBLEBUFFER,    // Flags
    PFD_TYPE_RGBA,        // The kind of framebuffer. RGBA or palette.
    32,                   // Colordepth of the framebuffer.
    0, 0, 0, 0, 0, 0,
    0,
    0,
    0,
    0, 0, 0, 0,
    24,                   // Number of bits for the depthbuffer
    0,                    // Number of bits for the stencilbuffer
    0,                    // Number of Aux buffers in the framebuffer.
    PFD_MAIN_PLANE,
    0,
    0, 0, 0
};
mPfd.cAlphaBits = 8;
```

This should unblock you for now.

**Comment by thirdmenson** [comment]
Thank you ! It works now. 
Is it possible to have the image based lighting on but just toggle off the environment visibility?

**Comment by romainguy** [comment]
What do you mean? You mean you want to not see the environment in the background? If so, yes. All you need is to not create a skybox.

**Comment by thirdmenson** [comment]
Yup that's what I meant and not setting the skyboxes works. Thanks ! Looks like I will be able to also toggle this on and off at runtime by setting skybox to null. Cheers !

**Comment by JonathanLehner** [comment]
How to do this with web? It seems the engine.createSwapChain() method does not accept any arguments

**Comment by prideout** [comment]
You can override the WebGL 2.0 context config in the 2nd arg to `Engine.create` as per https://google.github.io/filament/webgl/reference.html#Engine

**Comment by JonathanLehner** [comment]
like this?    this.engine = Filament.Engine.create(canvas, {alpha: true,premultipliedAlpha: false,preserveDrawingBuffer:true});
 ?

**Comment by JonathanLehner** [comment]
const options = {
  majorVersion: 2,
  minorVersion: 0,
  antialias: false,
  depth: true,
  alpha: true,
  premultipliedAlpha: false,
  preserveDrawingBuffer: true
};
this.engine = Filament.Engine.create(canvas, options);
I tried like this but it still doesn't seem to work

**Comment by as010101** [comment]
@bejado  
> So there's actually a pretty simple fix for this. In [`PlatformWGL.cpp`](https://github.com/google/filament/blob/master/filament/backend/src/opengl/PlatformWGL.cpp#L92) set `cAlphaBits` on the pixel descriptor:
> 
> ```
> mPfd = {
>     sizeof(PIXELFORMATDESCRIPTOR),
>     1,
>     PFD_DRAW_TO_WINDOW | PFD_SUPPORT_OPENGL | PFD_DOUBLEBUFFER,    // Flags
>     PFD_TYPE_RGBA,        // The kind of framebuffer. RGBA or palette.
>     32,                   // Colordepth of the framebuffer.
>     0, 0, 0, 0, 0, 0,
>     0,
>     0,
>     0,
>     0, 0, 0, 0,
>     24,                   // Number of bits for the depthbuffer
>     0,                    // Number of bits for the stencilbuffer
>     0,                    // Number of Aux buffers in the framebuffer.
>     PFD_MAIN_PLANE,
>     0,
>     0, 0, 0
> };
> mPfd.cAlphaBits = 8;
> ```
> 
> This should unblock you for now.

For some reason, I must use old filament version,  why I copy this code to `PlatformWGL.cpp` , run it on windows  ,  then get this error:
```
FEngine (64 bits) created at 000001EA91E3C080 (threading is enabled)
FEngine resolved backend: OpenGL
NVIDIA Corporation, NVIDIA GeForce RTX 3050 Laptop GPU/PCIe/SSE2, 4.1.0 NVIDIA 546.80, 4.10 NVIDIA via Cg compiler
OS version: 0
class utils::PostconditionPanic
in void __cdecl filament::PlatformWGL::makeCurrent(struct filament::backend::Platform::SwapChain *,struct filament::backend::Platform::SwapChain *) noexcept:241
in file D:\filament_build\new1\filament\filament\backend\src\opengl\PlatformWGL.cpp
reason: wglMakeCurrent() failed. hdc = FFFFFFFFC20144F5
```
